src/audio/copier/copier.c

```diff
@@ -416,50 +416,69 @@ static int create_dai(struct comp_dev *parent_dev, struct copier_data *cd,
 		break;
 	case ipc4_alh_link_output_class:
 	case ipc4_alh_link_input_class:
-		dai.type = SOF_DAI_INTEL_ALH;
-		dai.is_config_blob = true;
-		type = ipc4_gtw_alh;
-
-		/* copier
-		 * {
-		 *  gtw_cfg
-		 *  {
-		 *     gtw_node_id;
-		 *     config_length;
-		 *     config_data
-		 *     {
-		 *         count;
-		 *        {
-		 *           node_id;  \\ normal gtw id
-		 *           mask;
-		 *        }  mapping[MAX_ALH_COUNT];
-		 *     }
-		 *   }
-		 * }
-		 */
-		 /* get gtw node id in config data */
-		if (is_multi_gateway(node_id)) {
-			if (copier->gtw_cfg.config_length) {
-				const struct sof_alh_configuration_blob *alh_blob =
-					(const struct sof_alh_configuration_blob *)
-						copier->gtw_cfg.config_data;
-
-				dai_count = alh_blob->alh_cfg.count;
-				if (dai_count > IPC4_ALH_MAX_NUMBER_OF_GTW || dai_count < 0) {
-					comp_err(parent_dev, "Invalid dai_count: %d", dai_count);
+		{
+			dai.type = SOF_DAI_INTEL_ALH;
+			dai.is_config_blob = true;
+			type = ipc4_gtw_alh;
+			/* copier
+			 * {
+			 *  gtw_cfg
+			 *  {
+			 *     gtw_node_id;
+			 *     config_length;
+			 *     config_data
+			 *     {
+			 *         count;
+			 *        {
+			 *           node_id;  \\ normal gtw id
+			 *           mask;
+			 *        }  mapping[MAX_ALH_COUNT];
+			 *     }
+			 *   }
+			 *   tlv_type \\ required for ACE 2.0 and newer
+			 *   tlv_length
+			 *   dma_config
+			 *   {
+			 *       dma_method;
+			 *       pre_allocated_by_host;
+			 *       dma_channel_id;
+			 *       stream_id;
+			 *       dma_stream_channel_map;
+			 *   }
+			 * }
+			 */
+			/* get dma_config if exists */
+			int ret = ipc4_find_dma_config(&dai,
+							(uint8_t *)copier->gtw_cfg.config_data,
+							copier->gtw_cfg.config_length);
```

Comment:
we still have the problem that the OFLEN bit will not be set for SoundWire/ALH, so the firmware *cannot* program the  channel map in PCMSyCM registers.

It's only been 1.5 months that we talk about this.

Reply:
Yes, You are right. Actually reading of this field is left but data not utilized yet. It could change when EVAD will be supported, but for now I will remove that change.

Reply:
I think you can use the DMA stream_id but the device/channel mapping not really.

Reply:
Removed